pocs/camera/fli.py

```diff
@@ -0,0 +1,236 @@
+
+        if dark:
+            frame_type = libfli.FLI_FRAME_TYPE_DARK
+        else:
+            frame_type = libfli.FLI_FRAME_TYPE_NORMAL
+        self._FLIDriver.FLISetFrameType(self._handle, frame_type)
+
+        # For now set to 'visible' (i.e. light sensitive) area of image sensor.
+        # Can later use this for windowed exposures.
+        self._FLIDriver.FLISetImageArea(self._handle,
+                                        self._info['visible corners'][0],
+                                        self._info['visible corners'][1])
+
+        # No on chip binning for now.
+        self._FLIDriver.FLISetHBin(self._handle, bin_factor=1)
+        self._FLIDriver.FLISetVBin(self._handle, bin_factor=1)
+
+        # No pre-exposure image sensor flushing, either.
+        self._FLIDriver.FLISetNFlushes(self._handle, n_flushes=0)
+
+        # In principle can set bit depth here (16 or 8 bit) but most FLI cameras don't support it.
+        # Leave alone for now.
+
+        # Build FITS header
+        header = self._fits_header(seconds, dark)
+
+        # Start exposure
+        self._FLIDriver.FLIExposeFrame(self._handle)
+
+        # Start readout thread
+        exposure_event = Event()
+        readout_args = (filename,
+                        self._info['visible width'],
+                        self._info['visible height'],
+                        header,
+                        exposure_event)
+        readout_thread = Timer(interval=self._FLIDriver.FLIGetExposureStatus(self._handle).value,
+                               function=self._readout,
+                               args=readout_args)
+        readout_thread.start()
+
+        if blocking:
+            exposure_event.wait()
+
+        return exposure_event
+
+# Private Methods
+
+    def _readout(self, filename, width, height, header, exposure_event):
+
+        # Wait for exposure to complete. This should have a timeout in case something goes wrong.
+        while self._FLIDriver.FLIGetExposureStatus(self._handle) > 0 * u.second:
+            time.sleep(self._FLIDriver.FLIGetExposureStatus(self._handle).value)
+
+        # Readout.
+        # Use FLIGrabRow for now at least because I can't get FLIGrabFrame to work.
+        # image_data = self._FLIDriver.FLIGrabFrame(self._handle, width, height)
+        image_data = np.zeros((height, width), dtype=np.uint16)
+        for i in range(image_data.shape[0]):
+            try:
+                image_data[i] = self._FLIDriver.FLIGrabRow(self._handle, image_data.shape[1])
+            except RuntimeError as err:
+                self.logger.error('FLI camera readout error: expected {} rows, got {}!'.format(
+                    image_data.shape[0], i
+                ))
+                self.logger.error(err)
+                break
+
+        self._exposure_lock.release()
+
+        fits_utils.write_fits(image_data, header, filename, self.logger, exposure_event)
```

Comment:
Do you want to `write_fits` if you get the `RuntimeError` above?

Reply:
Arguable. The error could occur part way through readout, in which case you've still got part of an image that you might want to look at. Doesn't hurt to try to save regardless?

Reply:
Good point. Let's leave for now.